### ISSUE TITLE
Make flutter analyze configurable so it doesn't fail on info messages

### ISSUE THREAD

**Issue by CarlosZ** [flutter/flutter#31918]
## Use case

Possibly related to #20855, `flutter analyze` currently will error out for any message of any
severity detected by the linter/analyzer which is too strict when for example tightening analysis
options or deprecating members because it means that the problems must be fixed all
at once or ignored wholesale in `analysis_options.yaml` otherwise `flutter analyze` will fail.

## Proposal

Ideally `flutter analyze` could be configured to not fail on certain severities of messages likely using a command line flag.

**Comment by krisgiesing** [comment]
It looks like this would be a matter of adding a flag around here: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/commands/analyze.dart#L21

and then checking the error level here: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/commands/analyze_once.dart#L147

Should we take the comment at that line as a statement that the behavior requested here is undesirable, or just a statement about the current implementation?

**Comment by CarlosZ** [comment]
@krisgiesing It's unclear to me how we should take a comment. Maybe someone on the Flutter team can shed some light.

**Comment by krisgiesing** [comment]
@BondarenkoStas That question was for the Flutter team, not for @CarlosZ.

**Comment by MelbourneDeveloper** [comment]
@CarlosZ yes, this is a huge problem. We use the analyser in the pipelines and most rules should break the build, but we've found that there are a few rules that are mostly unavoidable sometimes. In these cases, we still want a warning because we don't want people needlessly breaking the rules, but we don't want it to break the build. Currently, we have to put `ignore` everywhere and it's pretty horrible. Info and warning should not break the build.

Theoretically, we should be able to stop fatal info and warning messages with `no-fatal-warnings`:

![image](https://user-images.githubusercontent.com/16697547/165643941-daff9502-b59a-4216-bf63-e7518965116c.png)

But, I changed this in our pipelines and the pipelines still fail when there is a warning. In other words, the problem seems to be that the `no-fatal-warnings` argument doesn't seem to work. As you can see here, it still errors out even though these are only warnings.

![image](https://user-images.githubusercontent.com/16697547/165644356-abdf6cee-e3e7-4e5f-98ad-5b971e76468e.png)

**Comment by emersonsiega** [comment]
> @CarlosZ @krisgiesing yes, this is a huge problem. We use the analyser in the pipelines and most rules should break the build, but we've found that there are a few rules that are mostly unavoidable sometimes. In these cases, we still want a warning because we don't want people needlessly breaking the rules, but we don't want it to break the build. Currently, we have to put `ignore` everywhere and it's pretty horrible. Info and warning should not break the build.
> 
> Theoretically, we should be able to stop fatal info and warning messages with `no-fatal-warnings`:
> 
> ![image](https://user-images.githubusercontent.com/16697547/165643941-daff9502-b59a-4216-bf63-e7518965116c.png)
> 
> But, I changed this in our pipelines and the pipelines still fail when there is a warning. In other words, the problem seems to be that the `no-fatal-warnings` argument doesn't seem to work. As you can see here, it still errors out even though these are only warnings.
> 
> ![image](https://user-images.githubusercontent.com/16697547/165644356-abdf6cee-e3e7-4e5f-98ad-5b971e76468e.png)



We have the exactly same necessity... We want to warn our devs about deprecated code, but without breaking the pipeline.

**Comment by MelbourneDeveloper** [comment]
I can second this @emersonsiega

The Dart analyzer has different behaviour from the flutter analyzer. I think the flutter analyzer is ignoring `no-fatal-warnings`, but the dart one doesn't

I have a feeling that the Flutter team doesn't want to distinguish between warnings and errors. The last time I looked at the analysis config, it is set up to fail the build on warnings.

**Comment by MelbourneDeveloper** [comment]
Incidentally, I was working with the Flutter repo recently...

I realized it doesn't seem to have any errors configured with error severity. They all seem to be warning only. And yet, if you get so much as a warning, it will break the build.

It doesn't really make sense. If an analysis rule will break the build, it should go red so we can see that it's going to break the build. It shouldn't be blue.

The whole point of error severity is that there is a sliding scale from red (break the build) down to blue (try not to do this, but if you really have to ...) and so on. Why doesn't the flutter repo take advantage of the error severities built into the Dart compiler?

**Comment by kevmoo** [comment]
Just hit this with a customer, too.

**Comment by MelbourneDeveloper** [comment]
> Just hit this with a customer, too. 

I don't understand why it doesn't just follow the same severity level config as the Dart analyser. 

There are flags to allow info and warnings not to break the build but the analyser ignores them.

**Comment by kevmoo** [comment]
Points to anyone who can create a PR here. I should be pretty straightforward to send things through to the Dart analyzer.

**Comment by MelbourneDeveloper** [comment]
> Points to anyone who can create a PR here. I should be pretty straightforward to send things through to the Dart analyzer.

I only use the Dart analyser for this reason.

But, I worry that I'm missing out on Flutter specific code violations

**Comment by kevmoo** [comment]
> > Points to anyone who can create a PR here. I should be pretty straightforward to send things through to the Dart analyzer.
> 
> I only use the Dart analyser for this reason.
> 
> But, I worry that I'm missing out on Flutter specific code violations

I don't think you have to worry here. There is nothing special in the flutter tool here.

**Comment by MelbourneDeveloper** [comment]
That's good to know @kevmoo. 

So, rules like [use_colored_box](https://dart-lang.github.io/linter/lints/use_colored_box.html) work with the standard Dart analyzer? 

If that's the case, why are there two different analyzers?

**Comment by kevmoo** [comment]
🤷 Not sure. @christopherfujino – do you know?

**Comment by christopherfujino** [comment]
> shrug Not sure. @christopherfujino – do you know?

I think you can run `dart analyze` on a flutter app...but you would have had to have already run `flutter pub get`. I suspect `flutter analyze` will ensure `flutter pub get` was run.

**Comment by chillbrodev** [comment]
> > shrug Not sure. @christopherfujino – do you know?
> 
> I think you can run `dart analyze` on a flutter app...but you would have had to have already run `flutter pub get`. I suspect `flutter analyze` will ensure `flutter pub get` was run.

You can indeed run `dart analyze` on a flutter app. It will run even if `flutter pub get` has not been ran. You are correct that `flutter analyze` will run `flutter pub get`. With that said, it is important to note that `dart analyze` has different commands. It only allows for `no-fatal-warnings` as noted by the help.

```
dart analyze --help
Analyze Dart code in a directory.

Usage: dart analyze [arguments] [<directory>]
-h, --help                   Print this usage information.
    --fatal-infos            Treat info level issues as fatal.
    --[no-]fatal-warnings    Treat warning level issues as fatal.
                             (defaults to on)

Run "dart help" to see global options.
```

**Comment by christopherfujino** [comment]
> You can indeed run `dart analyze` on a flutter app. It will run even if `flutter pub get` has not been ran. You are correct that `flutter analyze` will run `flutter pub get`. With that said, it is important to note that `dart analyze` has different commands. It only allows for `no-fatal-warnings` as noted by the help.

How would the analyzer be able to find the dependencies? I suspect if you tried this and analysis succeeded you might have the cached results of a previous flutter command present in a hidden `.dart_tool` directory.

**Comment by MelbourneDeveloper** [comment]
@christopherfujino is right. You need to run `flutter pub get`. If you run `flutter clean` and then run `dart analyze` straight afterward, it will not work.

Also, the Dart analyzer seems to work fine with Flutter specific lints/errors. I added the [austerity](https://pub.dev/packages/austerity) to a fresh Flutter app.

```yaml
include: package:austerity/analysis_options.yaml
```

This is the app
```dart
import 'package:flutter/material.dart';

void main() {
  runApp(
    MaterialApp(
      home: Scaffold(
        body: Container(
          color: Colors.red,
          child: const Text('Hello World!'),
        ),
      ),
    ),
  );
}
```

This is the output of `dart analyze`

> Analyzing flutter_application_17...    0.9s

>  error • lib/main.dart:7:15 • Use `ColoredBox`. • use_colored_box

> 1 issue found.

I.e. the dart analyzer can detect flutter specific rules.

This is good to know because this is what I use in GitHub actions and other pipelines. But, it makes me wonder why there is a flutter analyzer at all. It does have slightly differently options, but can't you adjust the dart analyzer to have the same sensitivity?

**Comment by christopherfujino** [comment]
> This is good to know because this is what I use in GitHub actions and other pipelines. But, it makes me wonder why there is a flutter analyzer at all. It does have slightly differently options, but can't you adjust the dart analyzer to have the same sensitivity?

Two non-trivial differences: flutter analyze has a watch mode, and it will run `flutter pub get` first.

**Comment by MelbourneDeveloper** [comment]
Brilliant. That explains a lot.

**Comment by christopherfujino** [comment]
> Brilliant. That explains a lot.

They also used to be more different, like it would also analyze your `example/` directory for plugins, but then this was upstreamed.

**Comment by MelbourneDeveloper** [comment]
@christopherfujino 

Still, it would be very nice if the Flutter analyser would honor the CLI arguments to not break on info or error.

**Comment by MostafaSolimanMO** [comment]
Any update on this?

**Comment by christopherfujino** [comment]
> Any update on this?

@MostafaSolimanMO Per https://github.com/flutter/flutter/wiki/Issue-hygiene#do-not-add-me-too-or-same-or-is-there-an-update-comments-to-bugs, please do not comment asking for updates on issues. You can influence the prioritization by adding a :+1: reaction to the original post.

**Comment by MelbourneDeveloper** [comment]
My workaround for this issue is to use the Dart analyzer in the pipeline. It seems to use all the Flutter lint rules. 

The Flutter analyzer is good when you're working with your IDE like vscode or Android Studio, but unfortunately, I can't get it to stop failing on warnings so it doesn't suit me in the pipeline. 

BTW: I turn nearly all warnings up to errors in the austerity package. That way, you can allow the occasional warning if you don't have time to fix all the warnings and gradually switch them over to error as you need.

**Comment by MuhammadAli-M** [comment]
As a workaround, I only filter the errors and warnings and invert the exit code:

```
        run: |
          if flutter analyze 2>&1 | grep -q -E 'error:|warning:'; 
          then exit 1
          else exit 0
          fi
```

**Comment by bkkoluch** [comment]
> My workaround for this issue is to use the Dart analyzer in the pipeline. It seems to use all the Flutter lint rules.
> 
> The Flutter analyzer is good when you're working with your IDE like vscode or Android Studio, but unfortunately, I can't get it to stop failing on warnings so it doesn't suit me in the pipeline.
> 
> BTW: I turn nearly all warnings up to errors in the austerity package. That way, you can allow the occasional warning if you don't have time to fix all the warnings and gradually switch them over to error as you need.

How did you make it work not to break the CICD?

I'm using codemagic and it's still breaking my builds with `dart analyze` command (having only info analysis hints).

**Comment by cupidchan** [comment]
This issue was opened bac in 2019. any words on getting this fixed?

**Comment by cupidchan** [comment]
> flutter analyze --no-fatal-infos --no-fatal-warnings

Thanks @Iarsbloch, I have just tried to run the exact command `flutter analyze --no-fatal-infos --no-fatal-warnings`. The message will still print out on screen but the return code is 0. Hence, we can rely on this to pass the test.
